src/NLog/Targets/Wrappers/ConcurrentRequestQueue.cs

```diff
@@ -0,0 +1,193 @@
+//   this list of conditions and the following disclaimer. 
+// 
+// * Redistributions in binary form must reproduce the above copyright notice,
+//   this list of conditions and the following disclaimer in the documentation
+//   and/or other materials provided with the distribution. 
+// 
+// * Neither the name of Jaroslaw Kowalski nor the names of its 
+//   contributors may be used to endorse or promote products derived from this
+//   software without specific prior written permission. 
+// 
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE 
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE 
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE 
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF 
+// THE POSSIBILITY OF SUCH DAMAGE.
+// 
+
+#if NET4_5 || NET4_0
+
+namespace NLog.Targets.Wrappers
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Collections.Generic;
+    using System.Threading;
+    using NLog.Common;
+
+    /// <summary>
+    /// Concurrent Asynchronous request queue based on <see cref="ConcurrentQueue{T}"/>
+    /// </summary>
+	class ConcurrentRequestQueue : IAsyncRequestQueue
+    {
+        private readonly ConcurrentQueue<AsyncLogEventInfo> _logEventInfoQueue = new ConcurrentQueue<AsyncLogEventInfo>();
+
+        /// <summary>
+        /// Initializes a new instance of the AsyncRequestQueue class.
+        /// </summary>
+        /// <param name="requestLimit">Request limit.</param>
+        /// <param name="overflowAction">The overflow action.</param>
+        public ConcurrentRequestQueue(int requestLimit, AsyncTargetWrapperOverflowAction overflowAction)
+        {
+            RequestLimit = requestLimit;
+            OnOverflow = overflowAction;
+        }
+
+        /// <summary>
+        /// Gets or sets the request limit.
+        /// </summary>
+        public int RequestLimit { get; set; }
+
+        /// <summary>
+        /// Gets or sets the action to be taken when there's no more room in
+        /// the queue and another request is enqueued.
+        /// </summary>
+        public AsyncTargetWrapperOverflowAction OnOverflow { get; set; }
+
+        public bool IsEmpty => _logEventInfoQueue.IsEmpty && Interlocked.Read(ref _count) == 0;
+
+        /// <summary>
+        /// Gets the number of requests currently in the queue.
+        /// </summary>
+        public int Count => (int)_count;
```

Comment:
Isn't this tricky?

Reply:
The Count is only there for debugging. IsEmpty is the one used for decisions.

Reply:
OK  clear, maybe we should set that in the `<remarks>`